### DISCUSSION TITLE
How can I get cshtml intellisense in a non-web project?

### DISCUSSION THREAD

**Discussion by VictorioBerra** [toddams/RazorLight#414] [Q&A] [answered]
I target SDK `Microsoft.NET.Sdk.Worker`, I created a cshtml and theres no intellisense inside it. I type `@model MyVM` and I get no completion when I do `@Model.Firstname` (or whatever MyVM member).

## csproj:

```xml
<Project Sdk="Microsoft.NET.Sdk.Worker">

  <PropertyGroup Label="Build">
    <TargetFramework>net5.0</TargetFramework>
    <UserSecretsId>redacted</UserSecretsId>
    <IsTransformWebConfigDisabled>true</IsTransformWebConfigDisabled>
  </PropertyGroup>

...


  <ItemGroup>
    <None Remove="Templates\GenericEmail.cshtml" />
  </ItemGroup>
  
  <ItemGroup>
    <Content Include="Templates\GenericEmail.cshtml" />
  </ItemGroup>


```

I have declared it as an embedded resource.

**Comment by jzabroski** [comment, marked as answer]
Use a transitive project reference.  Problem solved.

**Reply by VictorioBerra** [reply]
Thanks for your reply, this is my first ever discussion using the new GitHub discussions feature 🎉

Is there a .NET 5 specific feature for "transitive project reference" or do you mean the traditional sense in which I would:

1. Create a WebProject
2. Tell RazorLight engine to use UseEmbeddedResourcesProject for the assembly?

If so, without a direct dependency `MyApp -> MyRazorViews` MyApp wont know about my ViewModels.

Could you provide a link for me to read more or a super quick explanation on how I can achieve this?

**Reply by VictorioBerra** [reply]
@jzabroski 

- Or maybe you are saying `MyProject -> ViewModel Project -> Views Project (RazorLight and cshtml)`?
- Do I HAVE to target web Sdk to have intellisense?

Honestly I am struggling to understand what is meant by "Transitive Reference" especially compared to a direct reference, there are no simple examples on google for RazorLight and Core with cshtml files and .NET 5, and 90% of the blogs discussing RazorLight just use strings in their samples which seems like the wrong way to write HTML templates.

It would be really nice to have more samples in this repository.

**Reply by VictorioBerra** [reply]
Okay a little more info now:

Changing the above csproj to the following based on your samples lit up the intellisense

```xml
  <ItemGroup>
    <Compile Remove="Templates\**" />
    <None Remove="*.cshtml" />
    <EmbeddedResource Include="Templates\**" />
  </ItemGroup>
  ```
  
  However... I still want to hear about the transitive dependency solution. Maybe I should be putting this all in its own project?

**Reply by jzabroski** [reply]
I put my RazorLight code in a project called MyCompanyName.MyParentProjectName.Templating

I then reference it through a variety of "endpoint" projects that have system entrypoints.

**Reply by VictorioBerra** [reply]
If I follow this correctly, It sounds like you have:

`MyAPIApp`, `MyMVCApp`, and `MyConsoleApp` (these being the endpoints, or entrypoints) all reference `MyCompanyName.MyParentProjectName.Templating`. This situation would make it a **direct** dependency.

Otherwise, you might have `MyAPIApp`, `MyMVCApp`, and `MyConsoleApp` (these being the entrypoints) all reference `MyCompanyName.MyParentProjectName.Businesslogic` which ultimately references `MyCompanyName.MyParentProjectName.Templating`. This situation would make it a **transitive** dependency.

You also didn't clarify if `MyCompanyName.MyParentProjectName.Templating` targets web (`<Project Sdk="Microsoft.NET.Sdk.Web">` VS `<Project Sdk="Microsoft.NET.Sdk.Worker">`), I say this because you marked your comment as the answer but just making something a "transitive" dependency does not magically give it intellisense in VS.

**Reply by jzabroski** [reply]
Hi. Thanks for the discussion. By transitive dependency I am referring to the SDK, as opposed to Razorlight. Razorlight IntelliSense is not enabled by Razorlight as much as it is by the SDK.

So... it is a transitive dependency. It's just not obvious that when you type `Sdk="Microsoft.NET.Sdk.Web"` what you are partially saying is

```xml
<FrameworkReference Include="Microsoft.AspNet.App" />
```

The "magic" is at a couple of steps. 
1. Our `TemplatePageBase` exposes properties that emulate how Razor pages work
2. Your models are technically an instance of that template. They hang off the Model property.
3. Visual Studio needs the SDK to know to load the Razor designer. 
4. Specifying `@inherits` gives you intellisense on your viewmodel

Most people don't want an in depth answer so I try to keep it simple. Always willing to peel back layers.
